### ISSUE TITLE
[Feature request] router resolver should be cancellable

### ISSUE THREAD

**Issue by goldenbearkin** [angular/angular#21961]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

I search around the issues regarding this and did find one but it was closed (https://github.com/angular/angular/issues/13056). I believe it should rather be a feature request. It is completely legit to expect that the ongoing route and the observable sitting inside the resolve function should be discarded if the user changes their mind and preceded to another route. Youtube is an example.

## Current behavior

The scenario is the user clicks a slow route link that fetches data necessary for a component, and while that is resolving the user hits back, or clicks on another link. I would then want to cancel the current resolve and begin the new routing.

To demotrate: https://stackblitz.com/edit/angular-u6zquh

To try to solve the problem, I inject the router event into the SomeResolve class and hoping to do like... (below is pseudo code)
````ts
@Injectable()
export class SomeResolve implements Resolve<any> {

  constructor(private someService$: SomeService, private router$: Router) { }

  resolve() {
    return this.someService$
      .takeUntil(
        this.router$.events.pipe(
          filter(NavigationStart)
        )
      );
  }
}
````

 It seems that all the router events are blocked until resolver complete. Below is quoted from Angular.io router and navigation section.
> The Observable provided to the Router must complete. If the Observable does not complete, the navigation will not continue.

So, this is not a solution because no events are shooted before resolver function complete. User still needs to wait for the ajax/observable complete before proceeding.

## Expected behavior

The router events are not blocked or any idea which can lead to cancellable router resolver.

<pre><code>
Angular version: 5.2.0

**Comment by dereklin** [comment]
I am having the exact problem.  I have multiple routes in the app.  One lazily loaded route uses a resolver to get some async data.  When the network is slow, this route could take a minute to load.  So now I use the router lifecycle to show a spinner while the sync code is running.  But let's say the user clicks on this slow route and then wants to click on another route to get out, they are stuck at this slow route until the resolver resolves.

What should happen is that the slow resolver should be cancelled when route changes -- route change should take precedence of route resolver.

Route resolver is a nice idea and it's easy to use.  But when it's blocking, it's only use might be fore initial app bootstrapping.

For now, we have to manually write the observables in the component.

**Comment by jasonaden** [comment]
Thank you for this feature request, and yes, I agree. We are looking at creating some new APIs to make this work. I'll add the `feature` label here and we should be picking some of these up in the relatively near future.

**Comment by lukaszzieba** [comment]
Any progress?

**Comment by professphysics** [comment]
+1

**Comment by mym987** [comment]
While waiting for the official solution, this might be something interesting:

**Comment by mrusmanali** [comment]
Any progress?

**Comment by dereklin** [comment]
saw jason's talk on angular connect https://www.youtube.com/watch?v=2wMQTxtpvoY around 2:27:55.  not sure if it's in the latest beta.

**Comment by jadams88** [comment]
@dereklin I just watched the angular connect vid (thank you) and the [final slide](https://youtu.be/2wMQTxtpvoY?t=9136) shows that the updated router and Guard Priority will be available with Angular 7.2. 

However I'm currently using 7.1.4 and a quick play seems that the cancellable resolvers is already working along with  `urlUpdateStrategy: 'eager'`.

**Comment by atscott** [comment]
Closing, as I think this is outdated and no longer relevant. The router events are not blocked by pending resolvers and the observable should already be unsubscribed on a new navigation.
